### DISCUSSION TITLE
Shared memory, again. For core-to-core communication.

### DISCUSSION THREAD

**Discussion by kikuyuta** [eclipse-zenoh/roadmap#142] [Ideas and new features]
I am thinking it would be great if Zenoh could be used for communication between cores on a same die.

TL;DR
There are chips called SoC that have multiple cores. For example, TI AM3358 series and STM32MP1 series. They have rich CPUs such as A7/A8 and CPUs with limited resources such as M4. The rich CPUs can be used for GPOS, or General Purpose OS, such as Linux, and the poor CPUs can be used for real-time processing by RTOS.

However, communication between these CPUs is not easy, and building a system combining them requires engineers familiar with the very details of those chips.

Since these SoCs have shared memory, if Zenoh enables communication between CPUs on the SoC's shared memory,

- you can build a system without having to dive into the detailed specifications of the chips,
- Communication overhead is minimized and communication latency between CPUs can be optimized, and
- Putting a Zenoh router in a rich CPU allows the poor CPUs to communicate directly with external Zenoh nodes.

In the current zenoh version, zenoh-pico does not have a shared memory implementation. If the shared memory function becomes available in future versions, both in Rust and C versions, then core-to-core communication could be realized by typical programming style with Zenoh.

For example, there is OpenAMP for this purpose, and if Zenoh can replace it, I imagine it could become the de facto standard for core-to-core communication.
https://www.openampproject.org/
